File: kafka-connect-jdbc-sink/src/main/scala/com/datamountaineer/streamreactor/connect/jdbc/sink/JdbcSinkConnector.scala
package com.datamountaineer.streamreactor.connect.jdbc.sink

import java.util

import com.datamountaineer.streamreactor.connect.jdbc.sink.config.JdbcSinkConfig
import com.typesafe.scalalogging.slf4j.StrictLogging
import org.apache.kafka.connect.connector.Task
import org.apache.kafka.connect.errors.ConnectException
import org.apache.kafka.connect.sink.SinkConnector

import scala.collection.JavaConverters._
import scala.util.{Failure, Try}


/**
  * <h1>JdbcSinkConnector</h1>
  * Kafka connect Redis Sink connector
  *
  * Sets up RedisSinkTask and configurations for the tasks.
  **/
class JdbcSinkConnector extends SinkConnector with StrictLogging {
  //???
  private var configProps: util.Map[String, String] = null

  /**
    * States which SinkTask class to use
    **/
  override def taskClass(): Class[_ <: Task] = classOf[JdbcSinkTask]

  /**
    * Set the configuration for each work and determine the split
    *
    * @param maxTasks The max number of task workers be can spawn
    * @return a List of configuration properties per worker
    **/
  override def taskConfigs(maxTasks: Int): util.List[util.Map[String, String]] = {
    logger.info(s"Setting task configurations for $maxTasks workers.")
    (1 to maxTasks).map(c => configProps).toList.asJava
  }

  /**
    * Start the sink and set to configuration
    *
    * @param props A map of properties for the connector and worker
    **/
  override def start(props: util.Map[String, String]): Unit = {
    logger.info(s"Starting Hbase sink task with ${props.toString}.")
    configProps = props
    Try(new JdbcSinkConfig(props)) match {
      case Failure(f) => throw new ConnectException("Couldn't start Hbase Sink due to configuration error.", f)
      case _ =>
    }
  }

  override def stop(): Unit = {}

  override def version(): String = getClass.getPackage.getImplementationVersion
}

File: kafka-connect-jdbc-sink/src/main/java/com/datamountaineer/streamreactor/connect/jdbc/sink/JdbcSinkConnector.java
/**
 * Copyright 2015 Datamountaineer.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

package com.datamountaineer.streamreactor.connect.jdbc.sink;

import com.datamountaineer.streamreactor.connect.jdbc.sink.config.JdbcSinkConfig;
import com.google.common.base.Joiner;
import com.google.common.collect.Lists;
import org.apache.kafka.connect.connector.Task;
import org.apache.kafka.connect.errors.ConnectException;
import org.apache.kafka.connect.sink.SinkConnector;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;

/**
 * <h1>JdbcSinkConnector</h1>
 * Kafka connect JDBC Sink connector
 * <p>
 * Sets up JdbcSinkTask and configurations for the tasks.
 **/
public final class JdbcSinkConnector extends SinkConnector {
    private static final Logger logger = LoggerFactory.getLogger(JdbcSinkConnector.class);

    //???
    private Map<String, String> configProps = null;

    /**
     * States which SinkTask class to use
     **/
    public Class<? extends Task> taskClass() {
        return JdbcSinkTask.class;
    }

    /**
     * Set the configuration for each work and determine the split
     *
     * @param maxTasks The max number of task workers be can spawn
     * @return a List of configuration properties per worker
     **/
    @Override
    public List<Map<String, String>> taskConfigs(int maxTasks) {
        logger.info("Setting task configurations for " + maxTasks + " workers.");
        final List<Map<String, String>> configs = Lists.newArrayList();
        for (int i = 0; i < maxTasks; ++i)
            configs.add(configProps);
        return configs;
    }

    /**
     * Start the sink and set to configuration
     *
     * @param props A map of properties for the connector and worker
     **/
    @Override
    public void start(Map<String, String> props) {
        final Joiner.MapJoiner mapJoiner = Joiner.on(',').withKeyValueSeparator("=");
        final String propsStr = mapJoiner.join(props);
        logger.info("Starting JDBC sink task with " + propsStr);
        configProps = props;

        try {
            new JdbcSinkConfig(props);
        } catch (Throwable t) {
            throw new ConnectException("Couldn't start JDBC Sink due to configuration error.", t);
        }
    }

    @Override
    public void stop() {
    }

    @Override
    public String version() {
        return getClass().getPackage().getImplementationVersion();
    }
}
